File: LibreOJ/109/109.cpp
#include <bits/stdc++.h>

using namespace std;

const int mod = 998244353;

int n, m, op, u, v, fa[4000005];
long long ans;

int find(int x) {
    return x == fa[x] ? x : fa[x] = find(fa[x]);
}

void merge(int x, int y) {
    fa[find(x)] = find(y);
}

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 0; i <= n; i++) {
        fa[i] = i;
    }
    for (int i = 1; i <= m; i++) {
        scanf("%d%d%d", &op, &u, &v);
        if (op == 0) {
            merge(u, v);
        } else {
            ans <<= 1ll, ans |= 1ll * (find(u) == find(v));
        }
        ans %= mod;
    }
    printf("%lld\n", ans);
    return 0;
}

File: LibreOJ/109/109.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
    "strings"
)

const mod = 998244353

var fa [4000005]int

func main() {
    scanner := bufio.NewScanner(os.Stdin)

    scanner.Scan()
    raw := strings.Split(scanner.Text(), " ")
    n, _ := strconv.Atoi(raw[0])
    m, _ := strconv.Atoi(raw[1])

    for i := 1; i <= n; i++ {
        fa[i] = i
    }

    ans := 0

    for i := 1; i <= m; i++ {
        scanner.Scan()
        raw := strings.Split(scanner.Text(), " ")
        op, _ := strconv.Atoi(raw[0])
        u, _ := strconv.Atoi(raw[1])
        v, _ := strconv.Atoi(raw[2])

        if op == 0 {
            merge(u, v)
        } else { // op == 1
            ans <<= 1
            ans |= bool2int(find(u) == find(v))
            ans %= mod
        }
    }

    fmt.Println(ans)
}

func find(x int) int {
    if fa[x] == x {
        return x
    }

    fa[x] = find(fa[x])

    return fa[x]
}

func merge(x int, y int) {
    fa[find(x)] = find(y)
}

func bool2int(x bool) int {
    if x {
        return 1
    }

    return 0
}
